eth/watchers/pricefeedwatcher_test.go

```diff
@@ -58,20 +57,59 @@ func TestPriceFeedWatcher_UpdatePrice(t *testing.T) {
 	}
 }
 
-func TestPriceFeedWatcher_Watch(t *testing.T) {
+func TestPriceFeedWatcher_Subscribe(t *testing.T) {
 	require := require.New(t)
 	priceFeedMock := new(mockPriceFeedEthClient)
 	defer priceFeedMock.AssertExpectations(t)
 
-	w := &PriceFeedWatcher{
-		priceFeed:     priceFeedMock,
-		currencyBase:  "ETH",
-		currencyQuote: "USD",
+	w := &priceFeedWatcher{priceFeed: priceFeedMock}
+
+	// Start a bunch of subscriptions and make sure only 1 watch loop gets started
+	observedCancelWatch := []context.CancelFunc{}
+	cancelSub := []context.CancelFunc{}
+	for i := 0; i < 5; i++ {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		w.Subscribe(ctx, make(chan eth.PriceData, 1))
+
+		observedCancelWatch = append(observedCancelWatch, w.cancelWatch)
+		cancelSub = append(cancelSub, cancel)
+	}
+
+	require.NotNil(w.cancelWatch)
+	for i := range observedCancelWatch {
+		require.Equal(reflect.ValueOf(w.cancelWatch).Pointer(), reflect.ValueOf(observedCancelWatch[i]).Pointer())
 	}
 
+	// Stop all subscriptions and ensure watch loop stops running
+	for i := range cancelSub {
+		cancelSub[i]()
+		if i < len(cancelSub)-1 {
```

Comment:
1. Why only one subscription should be `nil`? 🤔 
2. Won't this `time.Sleep()` make this test flaky?

Reply:
1. The `cancelWatch` variable is used as an introspection here to know whether the watch loop is running or not. After unsubscribing the last subscription, the watch loop should stop and that variable is cleared as well, which is why the check changes for the last subscription. I rewrote this with some comments to make it clearer.
2. So this sleep is only waiting for the message from `cancelWatch()` to [end this loop](https://github.com/livepeer/go-livepeer/blob/cb39078434fa6600aed3131573ea90a71ae30e0a/eth/watchers/pricefeedwatcher.go#L145) and clear the `cancelWatch` variable (after calling it). It is an entirely "in memory" thing and given there is no I/O in these tests (the pricefeed client is mocked too), I find it extremely unlikely that that "message flow" is not going to complete in `1s` (the whole process would have to freeze I guess). I could try to make this sophisticated, in the tests (like a loop, checking multiple times) or in prod (have some other mechanism to propagate the change, like a `sync.Cond`), but had opted to keep it simple as I didn't think this could flake much. WDYT?

Reply:
Ok 👌